lib/jxl/decode.cc

```diff
@@ -451,6 +451,11 @@ struct JxlDecoderStruct {
   bool have_container;
   size_t box_count;
 
+  // The number of progressive steps in frame decoding.
+  // 0 = only after DC image ready
+  // 1 = also after full pass of group is ready
```

Comment:
This assumes single-frame images. What about multi-layer images (with zero-duration frames that get blended/combined)? Is the first step only when the DC of the final layer is available? It could make sense to show intermediate results, i.e. show the current blended image at each kRegularFrame (that would finally introduce an actual difference between kRegularFrame and kSkipProgressive).

Going from coarser to finer grained, I think this is the order:
- Show only fully decoded last/non-zero duration frames — i.e. not progressive at all, e.g. for loading a local file
- Show all fully decoded kRegularFrame frames, also the ones with zero duration (so multi-layer images get updated for every layer) — useful if we want to do things like layered encoding (e.g. starting with a subsampled layer), or something like a screenshot that is encoded with a modular layer for the nonphoto parts and one or more vardct frames for the photo parts
- Show DC images too (1:8) — useful for cheap basic progression
- Show full AC passes that are the last pass for their resolution target (so at most four: one for 1:8, one for 1:4, one for 1:2 and one for 1:1) — this would be useful when you don't need full resolution, so you can stop decoding e.g. when 1:2 is available
- Show all full AC passes — useful for general progressive decode, including saliency progression
- Show DC progressively too (e.g. 1:32, 1:16, only possible in case of dc frame) — useful to replace LQIP placeholders
- Show incomplete passes (trigger progression event whenever a group has been decoded) — useful when using e.g. center-first group ordering

The current value 0 is the third bullet and 1 is the fifth bullet.

I'm not saying we urgently need to implement all of these levels of granularity right now, but it would be good to take into account that more detail levels may be added later, and leave room for them.

Reply:
This reminds me: kSkipProgressive frames should skip emitting progressive events at all.

Reply:
> I added the 7 different stages of progressiveness you suggested to the enum. However in the api, we still expect only two different values. I don't know how to make this future-proof in the best way. Any suggestions?

Perhaps in the API you should be able to set a bitfield (as described by the ProgressiveDetail enum, but just as a uint32). Even when things are implemented, there will be no guarantee that any progressive event will be produced (e.g. single-group images will probably always just immediately do a full decode), so applications cannot expect that to happen anyway. So we could just allow setting any detail, and some of them will never actually make a difference (for now) but that's fine.

Reply:
Now I made an enum outside, (which does not contain `kNone`, since we signal "don't use progression" in another way).
I raise an error saying it is not implemented when it is not `kDC` or `kPasses` for now.

Reply:
kSkipProgressive check is added.